src/process/io/s3/S3Config.h

```diff
@@ -0,0 +1,332 @@
+/**
+ * Copyright (c) 2019 DKFZ - ODCF
+ *
+ * Distributed under the MIT License (license terms are at https://github.com/dkfz-odcf/FastqIndEx/blob/master/LICENSE.txt).
+ */
+
+#ifndef FASTQINDEX_S3CONFIG_H
+#define FASTQINDEX_S3CONFIG_H
+
+#include "common/ErrorAccumulator.h"
+#include "common/IOHelper.h"
+#include "process/io/s3/S3ServiceOptions.h"
+#include <aws/core/client/AWSClient.h>
+#include <aws/core/auth/AWSCredentials.h>
+#include <aws/core/client/ClientConfiguration.h>
+#include <experimental/filesystem>
+
+using namespace Aws;
+using namespace Aws::Client;
+using namespace Aws::Http;
+using namespace Aws::Utils;
+using namespace Aws::Utils::RateLimits;
+using namespace Aws::Utils::Threading;
+using namespace std;
+using namespace std::experimental::filesystem;
+
+class S3Config : public ErrorAccumulator {
+
+private:
+
+    Aws::Client::ClientConfiguration defaultConfiguration;
+
+    path requestedCredentialsFile;
+
+    path requestedConfigFile;
+
+    path usedCredentialsFile;
+
+    path usedConfigFile;
+
+    /**
+     * May be empty upon construction but is then set to default.
+     *
+     * Identifies a section within an ini file like
+     *
+     * [default]
+     * key = value
+     *
+     * [section 31]
+     * key = value
+     */
+    string selectedConfigSection;
+
+    /**
+     * Map of all loaded values
+     */
+    map<string, string> allValues;
+
+    /**
+     * If an error occurred during construction, this is stored here. Error messages are retrievable via
+     * getErrorMessages()
+     */
+    bool _isValid{false};
+
+    /**
+     * Called to resolve the credential and configuration file.
+     *
+     * Will set credentialsFile and configFile.
+     *
+     * @return true, if this went fine.
+     */
+    bool figureOutConfigurationFiles();
+
+protected:
+
+    /**
+     * This could be static but might be overridden by e.g. Test classes
+     * @return
+     */
+    virtual path getDefaultAWSCredentialsFile() {
```

Comment:
Now that you actually discriminate between a path (representing a filename) and the actual file, you may consider renaming this
```suggestion
    virtual path getDefaultAWSCredentialsPath() {
```
and the same for all other functions of this kind.